File: python/open_guardrail/guards/agent_permission.py
"""Validate agent actions against allowlist."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_DEFAULT_DENIED: list[str] = [
    "delete",
    "execute",
    "sudo",
    "rm -rf",
    "DROP TABLE",
    "shutdown",
    "reboot",
    "send email",
    "make payment",
    "transfer funds",
]


class _AgentPermission:
    def __init__(
        self,
        *,
        action: str = "block",
        allowed_actions: Optional[List[str]] = None,
        denied_actions: Optional[List[str]] = None,
    ) -> None:
        self.name = "agent-permission"
        self.action = action
        self._allowed = allowed_actions
        self._denied_patterns = [
            re.compile(re.escape(d), re.IGNORECASE)
            for d in (denied_actions or _DEFAULT_DENIED)
        ]

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        violations: list[str] = []

        for pat in self._denied_patterns:
            if pat.search(text):
                violations.append(pat.pattern)

        if self._allowed:
            allowed_pats = [
                re.compile(re.escape(a), re.IGNORECASE)
                for a in self._allowed
            ]
            has_allowed = any(
                p.search(text) for p in allowed_pats
            )
            if not has_allowed and not violations:
                violations.append("action_not_in_allowlist")

        triggered = len(violations) > 0
        score = min(len(violations) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="agent-permission",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Agent action not permitted"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "violations": violations,
                    "reason": (
                        "Text contains agent actions"
                        " that are denied or not in"
                        " the allowed actions list"
                    ),
                }
                if triggered
                else None
            ),
        )


def agent_permission(
    *,
    action: str = "block",
    allowed_actions: Optional[List[str]] = None,
    denied_actions: Optional[List[str]] = None,
) -> _AgentPermission:
    return _AgentPermission(
        action=action,
        allowed_actions=allowed_actions,
        denied_actions=denied_actions,
    )

File: packages/guards/src/agent-permission.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface AgentPermissionOptions {
  action: 'block' | 'warn';
  allowedActions?: string[];
  deniedActions?: string[];
}

const DEFAULT_ACTIONS: string[] = [
  'delete',
  'execute',
  'install',
  'sudo',
  'rm -rf',
  'drop table',
  'shutdown',
  'reboot',
  'format disk',
  'send email',
  'make payment',
  'transfer funds',
];

export function agentPermission(options: AgentPermissionOptions): Guard {
  const denied = (options.deniedActions ?? DEFAULT_ACTIONS).map((a) =>
    a.toLowerCase(),
  );
  const allowed = options.allowedActions?.map((a) => a.toLowerCase());

  return {
    name: 'agent-permission',
    version: '0.1.0',
    description: 'Validates agent actions against a permission allowlist',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      const found: string[] = [];

      for (const action of denied) {
        if (lower.includes(action)) {
          if (!allowed || !allowed.includes(action)) {
            found.push(action);
          }
        }
      }

      const triggered = found.length > 0;
      const score = triggered ? Math.min(found.length / 3, 1.0) : 0;

      return {
        guardName: 'agent-permission',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { deniedActionsFound: found } : undefined,
      };
    },
  };
}
